receiver/stdinreceiver/receiver.go

```diff
@@ -0,0 +1,133 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package stdinreceiver
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"os"
+	"sync"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.opentelemetry.io/collector/obsreport"
+	"go.uber.org/zap"
+)
+
+const (
+	transport = "stdin"
+	format    = "string"
+)
+
+var (
+	errNilNextLogsConsumer = errors.New("nil logsConsumer")
+	listenerEnabled        = sync.Once{}
+	listeners              []chan string
+	stdin                  = os.Stdin
+)
+
+// stdinReceiver implements the component.MetricsReceiver for stdin metric protocol.
+type stdinReceiver struct {
+	logger       *zap.Logger
+	config       *Config
+	logsConsumer consumer.LogsConsumer
+	shutdown     chan struct{}
+}
+
+// NewLogsReceiver creates the stdin receiver with the given configuration.
+func NewLogsReceiver(
+	logger *zap.Logger,
+	config Config,
+	nextConsumer consumer.LogsConsumer,
+) (component.LogsReceiver, error) {
+	if nextConsumer == nil {
+		return nil, errNilNextLogsConsumer
+	}
+
+	r := &stdinReceiver{
+		logger:       logger,
+		config:       &config,
+		logsConsumer: nextConsumer,
+	}
+
+	return r, nil
+}
+
+func startStdinListener(logger *zap.Logger) {
+	listenerEnabled.Do(func() {
+		reader := bufio.NewReader(stdin)
+		for {
+			scanner := bufio.NewScanner(reader)
+			scanner.Split(bufio.ScanLines) // Set up the split function.
+			for scanner.Scan() {
+				line := scanner.Text()
+				for _, listener := range listeners {
+					listener <- line
+				}
+			}
+			if err := scanner.Err(); err != nil {
+				logger.Error("Error reading stdin", zap.Error(err))
+			}
+		}
+	})
+}
+
+// Start starts the stdin receiver, adding it to the stdin listener.
+func (r *stdinReceiver) Start(_ context.Context, host component.Host) error {
+	//start listening to stdin - do this at most once.
+
+	r.shutdown = make(chan struct{})
+
+	listener := make(chan string)
+	listeners = append(listeners, listener)
+
+	go func() {
+		go startStdinListener(r.logger)
+		for {
+			select {
+			case nextLine := <-listener:
+				r.consumeLine(nextLine)
+			case <-r.shutdown:
+				return
+			}
+		}
+	}()
+
+	return nil
+}
+
+func (r *stdinReceiver) consumeLine(line string) {
+	ctx := obsreport.ReceiverContext(context.Background(), r.config.Name(), transport)
+	ctx = obsreport.StartLogsReceiveOp(ctx, r.config.Name(), transport)
+	ld := pdata.NewLogs()
+	rl := pdata.NewResourceLogs()
+	ld.ResourceLogs().Append(rl)
+	ill := pdata.NewInstrumentationLibraryLogs()
+	rl.InstrumentationLibraryLogs().Append(ill)
+	lr := pdata.NewLogRecord()
+	ill.Logs().Append(lr)
```

Comment:
What are your thoughts on batching in the receiver before pushing on down the pipeline?

Reply:
There is a batch processor meant to batch records, I'd go with that to avoid recreating a lot of code. Also since this is streaming, we never know when it will cut off.

Reply:
Sounds reasonable.